jemalloc-ctl/src/raw.rs

```diff
@@ -0,0 +1,244 @@
+//! Raw `malloctl` getter/setters
+
+use error::{cvt, Error, Result};
+use libc::c_char;
+use {mem, ptr, slice};
+
+/// Translates `name` to a `mib` (Management Information Base)
+///
+/// `mib`s are used to avoid repeated name lookups for applications that
+/// repeatedly query the same portion of `jemalloc`s `mallctl` namespace.
+///
+/// On success, `mib` contains an array of integers. It is possible to pass
+/// `mib` with a length smaller than the number of period-separated name
+/// components. This results in a partial MIB that can be used as the basis for
+/// constructing a complete MIB.
+///
+/// For name components that are integers (e.g. the `2` in `arenas.bin.2.size`),
+/// the corresponding MIB component will always be that integer. Therefore, it
+/// is legitimate to construct code like the following:
+///
+/// ```
+/// extern crate libc;
+/// extern crate jemallocator;
+/// extern crate jemalloc_ctl;
+///
+/// #[global_allocator]
+/// static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
+///
+/// fn main() {
+///     use jemalloc_ctl::raw;
+///     use libc::{c_uint, c_char};
+///     unsafe {
+///         let mut mib = [0; 4];
+///         let nbins: c_uint = raw::get(b"arenas.nbins\0").unwrap();
+///         raw::name_to_mib(b"arenas.bin.0.size\0", &mut mib).unwrap();
+///         for i in 0..4 {
+///             mib[2] = i;
+///             let bin_size: usize = raw::get_mib(&mut mib).unwrap();
+///             println!("arena bin {} has size {}", i, bin_size);
+///         }
+///     }
+/// }
+/// ```
+pub fn name_to_mib(name: &[u8], mib: &mut [usize]) -> Result<()> {
+    unsafe {
+        validate_name(name)?;
+
+        let mut len = mib.len();
+        cvt(jemalloc_sys::mallctlnametomib(
+            name as *const _ as *const c_char,
+            mib.as_mut_ptr(),
+            &mut len,
+        ))?;
+        assert_eq!(mib.len(), len);
+        Ok(())
+    }
+}
+
+/// Uses the MIB `mib` as key to the _MALLCTL NAMESPACE_ and reads its value.
+///
+/// The [`name_to_mib`] API translates a string of the key (e.g. `arenas.nbins`)
+/// to a `mib` (Management Information Base).
+pub fn get_mib<T: Copy>(mib: &[usize]) -> Result<T> {
```

Comment:
These are still unsafe since you can mismatch the type and end up with invalid values (e.g. null references).